core/src/main/java/org/elasticsearch/common/xcontent/XContentParserUtils.java

```diff
@@ -121,23 +121,26 @@ public static Object parseStoredFieldsValue(XContentParser parser) throws IOExce
      * @param parser      the current {@link XContentParser}
      * @param delimiter   the delimiter to use to splits the field's name
      * @param objectClass the object class of the object to parse
+     * @param objects     a list of objects the parsed one will be added to
      * @param <T>         the type of the object to parse
      * @return the parsed object
      * @throws IOException if anything went wrong during parsing or if the type or name cannot be derived
      *                     from the field's name
      */
-    public static <T> T parseTypedKeysObject(XContentParser parser, String delimiter, Class<T> objectClass) throws IOException {
+    public static <T> void parseTypedKeysObject(XContentParser parser, String delimiter, Class<T> objectClass, List<T> objects)
```

Comment:
I changed the signature to accept a consumer instead of a list, for the suggestions case it doesn't change much though. But I think its okay to have only one version for now. We can go back to a second variant if needed later.